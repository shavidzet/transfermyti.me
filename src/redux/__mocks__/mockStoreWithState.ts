import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
