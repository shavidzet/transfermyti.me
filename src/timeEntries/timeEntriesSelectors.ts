import { createSelector, Selector } from "reselect";
import * as R from "ramda";
import { activeWorkspaceIdSelector } from "~/workspaces/workspacesSelectors";
import {
  ReduxState,
  TimeEntryModel,
  TimeEntryTableViewModel,
} from "~/typeDefs";

export const sourceTimeEntriesSelector = createSelector(
  (state: ReduxState) => state.timeEntries.source,
  (timeEntriesById): TimeEntryModel[] => Object.values(timeEntriesById),
);

export const includedSourceTimeEntriesSelector = createSelector(
  sourceTimeEntriesSelector,
  (sourceTimeEntries): TimeEntryModel[] =>
    sourceTimeEntries.filter(sourceTimeEntry => sourceTimeEntry.isIncluded),
);

export const sourceTimeEntriesForTransferSelector = createSelector(
  includedSourceTimeEntriesSelector,
  sourceTimeEntries =>
    sourceTimeEntries.filter(sourceTimeEntry =>
      R.isNil(sourceTimeEntry.linkedId),
    ),
);

export const sourceTimeEntriesInActiveWorkspaceSelector = createSelector(
  sourceTimeEntriesSelector,
  activeWorkspaceIdSelector,
  (sourceTimeEntries, activeWorkspaceId) =>
    sourceTimeEntries.filter(
      sourceTimeEntry => sourceTimeEntry.workspaceId === activeWorkspaceId,
    ),
);

export const timeEntriesForInclusionsTableSelector = createSelector(
  (state: ReduxState) => state.allEntities.areExistsInTargetShown,
  sourceTimeEntriesInActiveWorkspaceSelector,
  (state: ReduxState) => state.projects.source,
  (state: ReduxState) => state.tasks.source,
  (
    areExistsInTargetShown,
    sourceTimeEntries,
    sourceProjectsById,
    sourceTasksById,
  ): TimeEntryTableViewModel[] =>
    sourceTimeEntries.reduce((acc, sourceTimeEntry) => {
      const existsInTarget = sourceTimeEntry.linkedId !== null;
      if (existsInTarget && !areExistsInTargetShown) {
        return acc;
      }

      const projectName = R.pathOr(
        null,
        [sourceTimeEntry.projectId, "name"],
        sourceProjectsById,
      );

      let taskName = null;
      if (sourceTimeEntry.taskId !== null) {
        taskName = R.pathOr(
          null,
          [sourceTimeEntry.taskId, "name"],
          sourceTasksById,
        );
      }

      return [
        ...acc,
        {
          ...sourceTimeEntry,
          projectName,
          taskName,
          existsInTarget,
          isActiveInSource: true,
          isActiveInTarget: existsInTarget,
        },
      ];
    }, [] as TimeEntryTableViewModel[]),
);

export const timeEntriesTotalCountsByTypeSelector = createSelector(
  timeEntriesForInclusionsTableSelector,
  timeEntriesForInclusionsTable =>
    timeEntriesForInclusionsTable.reduce(
      (acc, { existsInTarget, isIncluded }: TimeEntryTableViewModel) => ({
        existsInTarget: acc.existsInTarget + (existsInTarget ? 1 : 0),
        isIncluded: acc.isIncluded + (isIncluded ? 1 : 0),
      }),
      {
        existsInTarget: 0,
        isIncluded: 0,
      },
    ),
);

export const sourceTimeEntryCountByTagIdSelector = createSelector(
  sourceTimeEntriesSelector,
  sourceTimeEntries => {
    const timeEntryCountByTagIdField: Record<string, number> = {};

    for (const timeEntry of sourceTimeEntries) {
      for (const tagId of timeEntry.tagIds) {
        const currentValue = R.propOr<number, Record<string, number>, number>(
          0,
          tagId,
          timeEntryCountByTagIdField,
        );
        timeEntryCountByTagIdField[tagId] = currentValue + 1;
      }
    }

    return timeEntryCountByTagIdField;
  },
);

export const sourceTimeEntryCountByIdFieldSelectorFactory = (
  idField: string,
): Selector<ReduxState, Record<string, number>> =>
  createSelector(sourceTimeEntriesSelector, sourceTimeEntries => {
    const timeEntryCountByIdField: Record<string, number> = {};

    for (const timeEntry of sourceTimeEntries) {
      const parentId = timeEntry[idField];
      if (parentId) {
        const currentCountForId = R.propOr<
          number,
          Record<string, number>,
          number
        >(0, parentId, timeEntryCountByIdField);
        timeEntryCountByIdField[parentId] = currentCountForId + 1;
      }
    }

    return timeEntryCountByIdField;
  });
