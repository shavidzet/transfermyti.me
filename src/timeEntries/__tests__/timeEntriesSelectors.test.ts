import cases from "jest-in-case";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import * as timeEntriesSelectors from "../timeEntriesSelectors";

const TEST_STATE = {
  ...state,
  timeEntries: {
    source: {
      "8001": {
        id: "8001",
        description: "This is a test entry (01)",
        isBillable: false,
        start: "2019-06-25T18:00:00.000Z",
        end: "2019-06-25T23:00:00.000Z",
        year: 2019,
        isActive: false,
        clientId: "3001",
        projectId: "2001",
        tagIds: ["4001", "4002"],
        tagNames: ["tag-a", "tag-b"],
        taskId: "7001",
        userId: "6001",
        userGroupIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: "clock-entry-01",
        isIncluded: false,
        memberOf: "timeEntries",
      },
      "8002": {
        id: "8002",
        description: "This is a test entry (02)",
        isBillable: false,
        start: "2019-06-24T14:00:00.000Z",
        end: "2019-06-24T22:00:00.000Z",
        year: 2019,
        isActive: false,
        clientId: "3001",
        projectId: "2001",
        tagIds: [],
        tagNames: [],
        taskId: "7001",
        userId: "6001",
        userGroupIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: "clock-entry-02",
        isIncluded: false,
        memberOf: "timeEntries",
      },
      "8003": {
        id: "8003",
        description: "This is a test entry (03)",
        isBillable: false,
        start: "2019-06-23T14:30:00.000Z",
        end: "2019-06-24T04:30:00.000Z",
        year: 2019,
        isActive: false,
        clientId: "3001",
        projectId: "2001",
        tagIds: ["4003"],
        tagNames: ["tag-c"],
        taskId: "7002",
        userId: "6001",
        userGroupIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: "clock-entry-03",
        isIncluded: false,
        memberOf: "timeEntries",
      },
      "8004": {
        id: "8004",
        description: "This is a test entry (04)",
        isBillable: false,
        start: "2019-06-23T14:00:20.000Z",
        end: "2019-06-23T17:05:20.000Z",
        year: 2019,
        isActive: false,
        clientId: "3002",
        projectId: "2002",
        tagIds: [],
        tagNames: [],
        taskId: null,
        userId: "6001",
        userGroupIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
        memberOf: "timeEntries",
      },
      "8005": {
        id: "8005",
        description: "This is a test entry (05)",
        isBillable: false,
        start: "2019-06-22T14:00:19.000Z",
        end: "2019-06-22T19:15:19.000Z",
        year: 2019,
        isActive: false,
        clientId: "3002",
        projectId: "2002",
        tagIds: [],
        tagNames: [],
        taskId: null,
        userId: "6001",
        userGroupIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
        memberOf: "timeEntries",
      },
    },
    target: {
      "clock-entry-01": {
        id: "clock-entry-01",
        description: "This is a test entry (01)",
        isBillable: false,
        start: "2019-06-25T18:00:00.000Z",
        end: "2019-06-25T23:00:00.000Z",
        year: 2019,
        isActive: false,
        clientId: "clock-client-01",
        projectId: "clock-project-01",
        tagIds: ["clock-tag-01", "clock-tag-02"],
        tagNames: ["tag-a", "tag-b"],
        taskId: "clock-task-01",
        userId: "clock-user-01",
        userGroupIds: [],
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "8001",
        isIncluded: true,
        memberOf: "timeEntries",
      },
      "clock-entry-02": {
        id: "clock-entry-02",
        description: "This is a test entry (02)",
        isBillable: false,
        start: "2019-06-24T14:00:00.000Z",
        end: "2019-06-24T22:00:00.000Z",
        year: 2019,
        isActive: false,
        clientId: "clock-client-01",
        projectId: "clock-project-01",
        tagIds: [],
        tagNames: [],
        taskId: "clock-task-02",
        userId: "clock-user-01",
        userGroupIds: [],
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "8002",
        isIncluded: true,
        memberOf: "timeEntries",
      },
      "clock-entry-03": {
        id: "clock-entry-03",
        description: "This is a test entry (03)",
        isBillable: false,
        start: "2019-06-23T14:30:00.000Z",
        end: "2019-06-24T04:30:00.000Z",
        year: 2019,
        isActive: false,
        clientId: "clock-client-01",
        projectId: "clock-project-01",
        tagIds: ["clock-tag-01"],
        tagNames: ["tag-a"],
        taskId: "clock-task-03",
        userId: "clock-user-01",
        userGroupIds: [],
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "8003",
        isIncluded: true,
        memberOf: "timeEntries",
      },
    },
    isFetching: false,
  },
};

describe("within timeEntriesSelectors", () => {
  test("the sourceTimeEntriesSelector returns array of values from state.source", () => {
    const result = timeEntriesSelectors.sourceTimeEntriesSelector(TEST_STATE);

    expect(result).toEqual(Object.values(TEST_STATE.timeEntries.source));
  });

  cases(
    "the selectors match their snapshots",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceTimeEntriesSelector",
        selector: timeEntriesSelectors.includedSourceTimeEntriesSelector,
      },
      {
        name: "for the sourceTimeEntriesForTransferSelector",
        selector: timeEntriesSelectors.sourceTimeEntriesForTransferSelector,
      },
      {
        name: "for the sourceTimeEntriesInActiveWorkspaceSelector",
        selector:
          timeEntriesSelectors.sourceTimeEntriesInActiveWorkspaceSelector,
      },
      {
        name: "for the timeEntriesTotalCountsByTypeSelector",
        selector: timeEntriesSelectors.timeEntriesTotalCountsByTypeSelector,
      },
      {
        name: "for the sourceTimeEntryCountByTagIdSelector",
        selector: timeEntriesSelectors.sourceTimeEntryCountByTagIdSelector,
      },
    ],
  );

  cases(
    "the timeEntriesForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown",
    options => {
      const updatedState = {
        ...TEST_STATE,
        allEntities: {
          ...TEST_STATE.allEntities,
          areExistsInTargetShown: options.areExistsInTargetShown,
        },
      };
      const result = timeEntriesSelectors.timeEntriesForInclusionsTableSelector(
        updatedState,
      );

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ],
  );

  test("the sourceTimeEntryCountByIdFieldSelectorFactory matches its snapshot", () => {
    const getCounts = timeEntriesSelectors.sourceTimeEntryCountByIdFieldSelectorFactory(
      "6001",
    );
    const result = getCounts(TEST_STATE);

    expect(result).toMatchSnapshot();
  });
});
