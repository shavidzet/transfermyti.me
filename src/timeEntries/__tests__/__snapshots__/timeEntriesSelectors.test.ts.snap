// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within timeEntriesSelectors the selectors match their snapshots for the includedSourceTimeEntriesSelector 1`] = `
Array [
  Object {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": "2019-06-23T17:05:20.000Z",
    "entryCount": 0,
    "id": "8004",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": "2019-06-23T14:00:20.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": "2019-06-22T19:15:19.000Z",
    "entryCount": 0,
    "id": "8005",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": "2019-06-22T14:00:19.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the sourceTimeEntriesForTransferSelector 1`] = `
Array [
  Object {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": "2019-06-23T17:05:20.000Z",
    "entryCount": 0,
    "id": "8004",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": "2019-06-23T14:00:20.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": "2019-06-22T19:15:19.000Z",
    "entryCount": 0,
    "id": "8005",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": "2019-06-22T14:00:19.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the sourceTimeEntriesInActiveWorkspaceSelector 1`] = `
Array [
  Object {
    "clientId": "3001",
    "description": "This is a test entry (01)",
    "end": "2019-06-25T23:00:00.000Z",
    "entryCount": 0,
    "id": "8001",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-01",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "start": "2019-06-25T18:00:00.000Z",
    "tagIds": Array [
      "4001",
      "4002",
    ],
    "tagNames": Array [
      "tag-a",
      "tag-b",
    ],
    "taskId": "7001",
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3001",
    "description": "This is a test entry (02)",
    "end": "2019-06-24T22:00:00.000Z",
    "entryCount": 0,
    "id": "8002",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-02",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "start": "2019-06-24T14:00:00.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": "7001",
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3001",
    "description": "This is a test entry (03)",
    "end": "2019-06-24T04:30:00.000Z",
    "entryCount": 0,
    "id": "8003",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-03",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "start": "2019-06-23T14:30:00.000Z",
    "tagIds": Array [
      "4003",
    ],
    "tagNames": Array [
      "tag-c",
    ],
    "taskId": "7002",
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": "2019-06-23T17:05:20.000Z",
    "entryCount": 0,
    "id": "8004",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": "2019-06-23T14:00:20.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": "2019-06-22T19:15:19.000Z",
    "entryCount": 0,
    "id": "8005",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": "2019-06-22T14:00:19.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the sourceTimeEntryCountByTagIdSelector 1`] = `
Object {
  "4001": 1,
  "4002": 1,
  "4003": 1,
}
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the timeEntriesTotalCountsByTypeSelector 1`] = `
Object {
  "existsInTarget": 3,
  "isIncluded": 2,
}
`;

exports[`within timeEntriesSelectors the sourceTimeEntryCountByIdFieldSelectorFactory matches its snapshot 1`] = `Object {}`;

exports[`within timeEntriesSelectors the timeEntriesForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = false 1`] = `
Array [
  Object {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": "2019-06-23T17:05:20.000Z",
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8004",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": "2019-06-23T14:00:20.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": "2019-06-22T19:15:19.000Z",
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8005",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": "2019-06-22T14:00:19.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the timeEntriesForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = true 1`] = `
Array [
  Object {
    "clientId": "3001",
    "description": "This is a test entry (01)",
    "end": "2019-06-25T23:00:00.000Z",
    "entryCount": 0,
    "existsInTarget": true,
    "id": "8001",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-01",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": "2019-06-25T18:00:00.000Z",
    "tagIds": Array [
      "4001",
      "4002",
    ],
    "tagNames": Array [
      "tag-a",
      "tag-b",
    ],
    "taskId": "7001",
    "taskName": "Task 1",
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3001",
    "description": "This is a test entry (02)",
    "end": "2019-06-24T22:00:00.000Z",
    "entryCount": 0,
    "existsInTarget": true,
    "id": "8002",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-02",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": "2019-06-24T14:00:00.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": "7001",
    "taskName": "Task 1",
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3001",
    "description": "This is a test entry (03)",
    "end": "2019-06-24T04:30:00.000Z",
    "entryCount": 0,
    "existsInTarget": true,
    "id": "8003",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-03",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": "2019-06-23T14:30:00.000Z",
    "tagIds": Array [
      "4003",
    ],
    "tagNames": Array [
      "tag-c",
    ],
    "taskId": "7002",
    "taskName": "Task 2",
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": "2019-06-23T17:05:20.000Z",
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8004",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": "2019-06-23T14:00:20.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  Object {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": "2019-06-22T19:15:19.000Z",
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8005",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": "2019-06-22T14:00:19.000Z",
    "tagIds": Array [],
    "tagNames": Array [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": Array [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;
